Avoid ID clashes

cgzones opened the pull request:
Currently the the ID for a new paste is randomly generated in the caller of the database insert() function.  Then the insert() function tries to insert a new row into the database with that passed ID.  There can however already exists a paste in the database with the same ID leading to an insert failure, due to a constraint violation due to the PRIMARY KEY attribute.  Checking prior the the INSERT via a SELECT query would open the window for a race condition.

A failure to push a new paste is quite severe, since the user might have spent some some to format the input.

Generate the ID in a loop inside, until the INSERT succeeds.

matze commented:
An ID clash is indeed an extreme issue but perhaps we can solve it differently by constructing an `INSERT` query that generates a random primary key?

cgzones commented:
There exists a sqlite built-in function[`random`](https://www.sqlite.org/lang_corefunc.html#random)(and [`randomblob`](https://www.sqlite.org/lang_corefunc.html#randomblob)), but this would not solve this issue, cause the function might still generate a result already existent in the database.
Also we need the ID for the URL path generation, so this approach would require a secondary SQL query to get the ID in all cases (also non-clashing).

cgzones committed:
Avoid ID clash

Currently the the ID for a new paste is randomly generated in the caller
of the database insert() function.  Then the insert() function tries to
insert a new row into the database with that passed ID.  There can
however already exists a paste in the database with the same ID leading
to an insert failure, due to a constraint violation due to the PRIMARY
KEY attribute.  Checking prior the the INSERT via a SELECT query would
open the window for a race condition.

A failure to push a new paste is quite severe, since the user might have
spent some some to format the input.

Generate the ID in a loop inside, until the INSERT succeeds.

matze commented on crates/wastebin_core/src/db.rs:
```diff
@@ -447,31 +446,56 @@ impl Handler {
+                    && extended_code == 1555
+                    && counter < 10 =>
+                {
+                    /* Retry if ID is already existent */
```
Please change to standard Rust comments with `//`.

cgzones replied:
done

matze commented on crates/wastebin_core/src/db.rs:
```diff
@@ -447,31 +446,56 @@ impl Handler {
+                    },
+                    Some(ref _message),
+                )) if code == rusqlite::ErrorCode::ConstraintViolation
+                    && extended_code == 1555
```
Can you add a link to some documentation?

cgzones replied:
done
